Shard eval dataset and aggregate eval metrics

jysohn23 opened the pull request:
Also, instead of calling `eval_loss.item()` every time do summation with
tensors on device.

jysohn23 committed:
Shard eval dataset and aggregate eval metrics

Also, instead of calling `eval_loss.item()` every time do summation with
tensors on device.

taylanbil commented on examples/run_glue_tpu.py:
```diff
@@ -256,14 +257,29 @@ def evaluate(args, model, tokenizer, prefix="", disable_logging=False):
         results.update(result)
+        results['eval_loss'] = eval_loss.item()
+
+        # Average all metrics from each shard
```
what are some of the metrics? does it make sense to avg all of them? some metrics may be additive.

jysohn23 replied:
f1, accuracy, eval_loss, acc_and_f1 (avg of the two)

I checked they all make sense averaged.

jysohn23 replied:
Yep ignore that :D Not additive as discussed. Will update PR.

taylanbil replied:
I'd be surprised if `global f1 == np.mean(local f1s)`. That's probably not true, let's verify on paper what the true formula to get global f1 is.

jysohn23 replied:
Updated to sync the pred/label tensors directly instead (these shouldn't be that big for finetuning tasks; single integer 0/1 per example). This way we don't have to have some custom aggregation per metric and don't touch upstream core code.

taylanbil commented on examples/run_glue_tpu.py:
```diff
@@ -256,14 +257,29 @@ def evaluate(args, model, tokenizer, prefix="", disable_logging=False):
-                logger.info("  %s = %s", key, str(result[key]))
-                writer.write("%s = %s\n" % (key, str(result[key])))
-                tb_writer.add_scalar(key, result[key])
+        if xm.is_master_ordinal():
```
is everything being logged here already on cpu?

jysohn23 replied:
Yes.

taylanbil replied:
maybe let's add a comment? It's a subtle point that can be missed by code readers.

taylanbil commented on examples/run_glue_tpu.py:
```diff
@@ -256,14 +257,29 @@ def evaluate(args, model, tokenizer, prefix="", disable_logging=False):
+
+        # Average all metrics from each shard
+        def reduce_fn(results):
+            avg_result = defaultdict(int)
```
why not `defaultdict(float)`?

jysohn23 replied:
Good point. Done.

taylanbil reviewed (commented):
did you test this e2e?

jysohn23 committed:
Change defaultdict to float

jysohn23 reviewed (commented):
Yeah, e2e tested.

jysohn23 committed:
Reduce the pred, label tensors instead of metrics

As brought up during review some metrics like f1 cannot be aggregated
via averaging. GLUE task metrics depends largely on the dataset, so
instead we sync the prediction and label tensors so that the metrics can
be computed accurately on those instead.
